### DISCUSSION TITLE
Real-time I2S audio mixing/passthrough?

### DISCUSSION THREAD

**Discussion by byoungq** [earlephilhower/arduino-pico#2210] [Q&A] [answered]
I'm attempting to salvage a project of mine by setting up a Pico W as a real time I2S audio mixer. After reading through the wiki and examples, I was able to write a short program to receive two standard I2S inputs and output a mixed one, When I tested it however, the output was all noise, almost like a jet turbine. The input devices are configured as I2S slave and the output device unfortunately must be configured as I2S master. I'll put my code below - Any advice would be greatly appreciated. (As an alternative question, is this even feasible with an rp2040?)

```
#include <I2S.h>

#define i2sMic1BclkPin 10 // GPIO pin numbers
#define i2sMic1LrclkPin 11
#define i2sMic1DoutPin 12

#define i2sMic2BclkPin 14 // GPIO pin numbers
#define i2sMic2LrclkPin 15
#define i2sMic2DoutPin 16

#define i2soutputBclkPin 19 // GPIO pin numbers
#define i2soutputLrclkPin 20
#define i2soutputDinPin 21

I2S i2s1(INPUT);
I2S i2s2(INPUT);
I2S i2s3(OUTPUT);

void setup() {
  Serial.begin(115200);
  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, 1);

  // i2s1.setSysClk(48000);
  // i2s2.setSysClk(48000);
  // i2s3.setSysClk(48000);

  // Set up I2S pins for inputs
  i2s1.setBCLK(i2sMic1BclkPin);
  i2s1.setDATA(i2sMic1DoutPin);
  i2s1.setBitsPerSample(16);
  i2s1.setBuffers(4, 128);

  
  // Set up I2S pins for inputs
  i2s2.setBCLK(i2sMic2BclkPin);
  i2s2.setDATA(i2sMic2DoutPin);
  i2s2.setBitsPerSample(16);
  i2s2.setBuffers(4, 128);
  
  // Set up I2S pins for output
  i2s3.setBCLK(i2soutputBclkPin);
  i2s3.setDATA(i2soutputDinPin);
  i2s3.setBitsPerSample(16);
  i2s3.setBuffers(4, 128);
  
  // Initialize I2S for inputs
  if (!i2s1.begin(48000)) {
    Serial.println("Failed to initialize I2S for inputs");
    while (1); // do nothing
  }
  
  // Initialize I2S for output
  if (!i2s2.begin(48000)) {
    Serial.println("Failed to initialize I2S for output");
    while (1); // do nothing
  }
  
  if (!i2s3.begin(48000)) {
    Serial.println("Failed to initialize I2S!");
    while (1); // do nothing
  }
}

void loop() {
  int16_t sample1Left, sample1Right, sample2Left, sample2Right;
  int16_t mixedLeft, mixedRight;

  // Read samples from input1
  if (i2s1.read16(&sample1Left, &sample1Right)) {
    //Passthrough test
    i2s3.write16(sample1Left, sample1Right);

    /*// Disabled mixing code
    //Read samples from input2
    if (i2s2.read16(&sample2Left, &sample2Right)) {
      // Mix samples
      mixedLeft = (sample1Left + sample2Left) / 2;
      mixedRight = (sample1Right + sample2Right) / 2;

      Write mixed samples to output
      i2s3.write16(mixedLeft, mixedRight);
    }
    */
  }
}
```

**Comment by Andy2No** [comment]
One thing that occurs to me is that the addition can cause an overflow, though that shouldn't explain always just getting noise because it wouldn't happen for smaller signals.  You could eliminate that as a problem by doing the division first:

(a + b) / 2 can overflow
(a/2 + b/2) can't, but does lose some precision.

Also, what happens if you lower the bit rate significantly?  I don't know if there's a minimum but maybe try 12kHz first, which ought to give better bandwidth than a landline telephone, enough to reproduce speech.

**Comment by earlephilhower** [comment, marked as answer]
> ...The input devices are configured as I2S slave and the output device unfortunately must be configured as I2S master...

There is no I2S slave mode support in the core.  You can use those input devices since the Pico will be the master, but the output one needs the Pico to be a slave and it's not going to work at all.

**Reply by byoungq** [reply]
That's what I was worried about - thanks for the help anyways.
I think I'll do a redesign with audio codecs so the mixing can be done analog rather than digital.

Thanks for writing the core, it makes working with the pico a whole lot easier.
